### PR TITLE
Refactor shell scripts

### PR DESCRIPTION
**Pull request by stefaniuk**
## Description

This is the second part of the refactoring effort to ensure that the shell scripts:

- Follow a consistent convention,
- Pass the ShellCheck linting test (please run `make shellscript-lint-all`), and
- Work either with the pre-installed CLI tooling or run commands using Docker containers.

## Context

Information for a reviewer on how to test the scripts with the `scan-secrets.sh` as an example:

Install the CLI tool

```
# On macOS
brew install gitleaks
# On Ubuntu
apt install gitleaks
```

Run a test

```
./scripts/githooks/scan-secrets.sh
VERBOSE=1 ./scripts/githooks/scan-secrets.sh # Use the gitleaks CLI tool if installed
VERBOSE=1 FORCE_USE_DOCKER=1 ./scripts/githooks/scan-secrets.sh # Use Docker
```

## Type of changes

- [x] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

**Review by ballr73 — APPROVED** [review]

File: scripts/githooks/check-file-format.sh [hunk 1]
```diff
@@ -70,13 +67,42 @@ function main() {
...
+    filter="$filter" dry_run_opt="${dry_run_opt:-}" cli-run-editorconfig
+  else
+    filter="$filter" dry_run_opt="${dry_run_opt:-}" docker-run-editorconfig
+  fi
+}
+
+# Run editorconfig natively.
+# Arguments (provided as environment variables):
+#   dry_run_opt=[dry run option]
+#   filter=[filter for files to check]
```
Comment (amaanibn-nasar1-nhs) [inline scripts/githooks/check-file-format.sh:80]: I think these comments could be removed if the things they are describing are more clear
cli-run-editorconfig() -> native-run-editorconfig() _(or something similar)_

File: scripts/githooks/check-markdown-format.sh [hunk 2]
```diff
@@ -53,15 +51,43 @@ function main() {
...
+      files="$files" cli-run-markdownlint
+    else
+      files="$files" docker-run-markdownlint
+    fi
   fi
 }
 
+# Run markdownlint natively.
+# Arguments (provided as environment variables):
+#   files=[files to check]
```
Comment (amaanibn-nasar1-nhs) [inline scripts/githooks/check-markdown-format.sh:64]: Same comment as above

File: scripts/reports/cloc-repository.sh [hunk 3]
```diff
@@ -28,20 +26,44 @@ function main() {
   enrich-report
 }
 
+# Create the report.
 function create-report() {
```
Comment (amaanibn-nasar1-nhs) [inline scripts/reports/cloc-repository.sh:30]: Could this be renamed to be create-lines-of-code-report?

File: scripts/reports/generate-sbom.sh [hunk 4]
```diff
@@ -29,16 +27,41 @@ function main() {
   enrich-report
 }
 
+# Create the report.
 function create-report() {
```
Comment (amaanibn-nasar1-nhs) [inline scripts/reports/generate-sbom.sh:31]: Could this be renamed to be create-sbom-report?

**Review by amaanibn-nasar1-nhs — COMMENTED** [review]
A few questions

File: scripts/githooks/check-file-format.sh [hunk 5]
```diff
@@ -70,13 +67,42 @@ function main() {
       ;;
   esac
 
+  if command -v editorconfig > /dev/null 2>&1 && ! is-arg-true "${FORCE_USE_DOCKER:-false}"; then
+    filter="$filter" dry_run_opt="${dry_run_opt:-}" cli-run-editorconfig
+  else
+    filter="$filter" dry_run_opt="${dry_run_opt:-}" docker-run-editorconfig
```
[the remark this answers is not in the document]
Comment (stefaniuk) [reply scripts/githooks/check-file-format.sh:73]: This functionality switches the `editorconfig` to the dry run mode by adding `--dry-run` flag to the command execution. This should output the files to be checked without running the check itself. I moved both, the `check` and `dry_run` variables to the `Options` section in the header for clarity. Thanks

File: scripts/githooks/check-file-format.sh (hunk 1 above)
Comment (stefaniuk) [reply scripts/githooks/check-file-format.sh:80]: This is the header of the function. I suggest we keep it as it communicates the dependency on the variables needed to be set for the function to run correctly. However, I'll rename this function (and others) for clarity as recommended. Thanks.

e.g.
- `run-editorconfig-natively`
- `run-editorconfig-in-docker`

Comment (stefaniuk) [reply scripts/githooks/check-file-format.sh:80]: I removed the headers across the codebase where there are no arguments to be passed, and the name of the function clearly expresses its intent.

File: scripts/githooks/check-markdown-format.sh (hunk 2 above)
Comment (stefaniuk) [reply scripts/githooks/check-markdown-format.sh:64]: cool, thanks. see the above comment

File: scripts/reports/generate-sbom.sh (hunk 4 above)
Comment (stefaniuk) [reply scripts/reports/generate-sbom.sh:31]: done, thanks

File: scripts/reports/cloc-repository.sh (hunk 3 above)
Comment (stefaniuk) [reply scripts/reports/cloc-repository.sh:30]: done, thanks

File: scripts/githooks/check-file-format.sh [hunk 6]
```diff
@@ -6,13 +6,16 @@ set -euo pipefail
 
 # Pre-commit git hook to check the EditorConfig rules compliance over changed
 # files. It ensures all non-binary files across the codebase are formatted
-# according to the style defined in the `.editorconfig` file.
+# according to the style defined in the `.editorconfig` file. This is a
+# editorconfig command wrapper. It will run editorconfig natively if it is
+# installed, otherwise it will run it in a Docker container.
```
[the remark this answers is not in the document]
Comment (stefaniuk) [reply scripts/githooks/check-file-format.sh:11]: This is by design, to allow customisation of the tooling without having to change any file in the repository template. If an engineering team would aim at strong process consistency, they can run `export FORCE_USE_DOCKER=true` to ensure it. As CI/CD pipeline runners are ephemeral, this behaviour is guaranteed, and the expected versions as configured will be run. I added a [note](https://github.com/nhs-england-tools/repository-template/pull/141/commits/cd2e267b27a6739a7585cd0648600968e00c2e7a#diff-10a959c7a1a531dc0fe97c6b4db94f8387242b197cba739ec221d44ee5e57033R143) on it.

File: scripts/reports/cloc-repository.sh (hunk 3 above)
Comment (stefaniuk) [reply scripts/reports/cloc-repository.sh]: I changed the file name for clarity.

File: scripts/reports/generate-sbom.sh (hunk 4 above)
Comment (stefaniuk) [reply scripts/reports/generate-sbom.sh:30]: I changed the file name for readability and clarity.

**Review by jacobgill1 — APPROVED** [review]
LGTM

**Review by amaanibn-nasar1-nhs — APPROVED** [review]
LGTM

**Review by regularfry — APPROVED** [review]

**Pull request merged**